rclcpp_components/src/node_main.cpp

```diff
@@ -0,0 +1,68 @@
+// Copyright 2016 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "class_loader/class_loader.hpp"
+#include "rclcpp/rclcpp.hpp"
+#include "rclcpp_components/node_factory.hpp"
+
+
+#define LINKTIME_COMPOSITION_LOGGER_NAME "node_main"
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  rclcpp::init(argc, argv);
+  rclcpp::Logger logger = rclcpp::get_logger(LINKTIME_COMPOSITION_LOGGER_NAME);
+  rclcpp::executors::SingleThreadedExecutor exec;
+  rclcpp::NodeOptions options;
```

Comment:
I think we should populate NodeOptions with any non-ROS arguments. We can get them by replace the `init()` above with `init_and_remove_ros_arguments()`:

```c++
auto args = rclcpp::init_and_remove_ros_arguments(argc, argv);
...
rclcpp::NodeOptions options;
options.arguments(args);
```

Reply:
Doesn't this happen automatically, so long as http://docs.ros2.org/dashing/api/rclcpp/classrclcpp_1_1NodeOptions.html#a716a84c68dc2307b03171cc1a8663549 is not explicitly set to `false`?

Reply:
I think I checked with a simple example and `options.arguments` was always empty, but maybe it's worth checking again.

Reply:
They will be empty, but the context has the global arguments. What do you want them to be used for? Is that just for the user to access the command line arguments or for the node to use them?

Reply:
[For example](https://github.com/ros2/demos/blob/8d26b6154f6740430f460ead375425d387900657/image_tools/src/cam2image.cpp#L75), when the node is created we want to access the command line arguments to configure the node.

Reply:
I see, but I'm not sure this is the right thing to do. But I'm also not sure it's the wrong thing to do. I would have thought the node would get the arguments from the context, as they are already passed to `rclcpp::init` (essentially `rclcpp::Context::init`).

Really, I think we need a method to "get the command line arguments" which looks at both the global arguments from the context and the arguments given by the node options and either picks one of them or combines them or something. Maybe respecting the `use_global_arguments` option in the node options.

I suppose we can leave this as-is for now, but the story for a node class accessing CLI arguments intended for it is not that clear to me yet.